protocols/autonatv2/src/server/behaviour.rs

```diff
@@ -0,0 +1,235 @@
+        remote_addr: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(Either::Right(dial_request::Handler::new(
+            remote_addr.clone(),
+            self.rng.clone(),
+        )))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        _role_override: Endpoint,
+        port_use: PortUse,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if port_use == PortUse::New {
+            self.handlers.insert(
+                (
+                    addr.iter()
+                        .filter(|e| !matches!(e, Protocol::P2p(_)))
+                        .collect(),
+                    peer,
+                ),
+                connection_id,
+            );
+        }
+        Ok(Either::Left(dial_back::Handler::new()))
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            FromSwarm::DialFailure(DialFailure {
+                error: DialError::WrongPeerId { .. },
+                peer_id: Some(peer_id),
+                ..
+            }) => {
+                if let Some(key) = self
+                    .dialing_dial_back
+                    .keys()
+                    .find(|(_, p)| *p == peer_id)
+                    .cloned()
+                {
+                    let cmds = self.dialing_dial_back.remove(&key).unwrap();
+                    for cmd in cmds {
+                        let _ = cmd.back_channel.send(DialBack::Dial);
+                    }
+                }
+            }
+            FromSwarm::DialFailure(DialFailure {
+                error: DialError::Transport(pairs),
+                ..
+            }) => {
+                for (addr, _) in pairs.iter() {
+                    let cleaned_addr: Multiaddr = addr
+                        .iter()
+                        .filter(|p| !matches!(p, Protocol::P2p(_)))
+                        .collect();
+                    let peer_id_opt = addr.iter().find_map(|p| match p {
+                        Protocol::P2p(peer) => Some(peer),
+                        _ => None,
+                    });
+                    if let Some(peer_id) = peer_id_opt {
+                        if let Some(cmd) = self.dialing_dial_back.remove(&(cleaned_addr, peer_id)) {
+                            for cmd in cmd {
+                                let _ = cmd.back_channel.send(DialBack::Dial);
+                            }
+                        }
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        _connection_id: ConnectionId,
+        event: <Handler<R> as ConnectionHandler>::ToBehaviour,
+    ) {
+        if let Either::Right(m) = event {
+            match m {
```

Comment:
Please prefer "deep" pattern matching over nested `if let`s. What we want to optimise for is minimal indentation (makes code easier to read). In this case, where the majority of code is within one match arm, we can make that the main flow of the function. At the same time, I'd like us to not use `else` where possible. Instead, use control flow to early-exit.

```rust
let cmd = match event {
	Either::Right(Ok(cmd)) => cmd,
	Either::Right(Err(e) => {
		tracing::warn!("incoming dial request failed: {}", e);
		return	
	};
	Either::Left(_) => {
		return;
	};
};

if let Some(connection_id) = self.handlers.get(&(addr.clone(), peer_id)) {
	self.pending_events.push_back(ToSwarm::NotifyHandler {
		peer_id,
		handler: NotifyHandler::One(*connection_id),
		event: Either::Left(cmd),
	});
	return;
}

if let Some(pending) = self.dialing_dial_back.get_mut(&(addr.clone(), peer_id)) {
	pending.push_back(cmd);
	return;
}

self.pending_events.push_back(ToSwarm::Dial {
    opts: DialOpts::peer_id(peer_id)
        .addresses(Vec::from([addr.clone()]))
        .condition(PeerCondition::Always)
        .allocate_new_port()
        .build(),
});
self.dialing_dial_back.insert((addr, peer_id), VecDeque::from([cmd]));
```

Reply:
That's a left over as well. I have changed it. It's one more nesting then you requested, but still